docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,555 @@
+`MeshHTTPRoute`.
+
+#### Merging behavior
+
+With `targetRef` resources, the ultimate configuration is a result of overriden/merged policies depending on their specificity.
+
+> :question: Is the current merge behavior what we need for routing policy? What surprises are there? See the next sections
+
+##### Service owner routes
+
+There's a discussion in the Gateway API repository about what different types of
+routes exist and how they should interact. Consider:
+
+- the owner of a service `svc_a`
+may want to configure a routing policy for their service that applies to requests from all services.
+- another service owner may want to apply routing to requests from their service
+to `svc_a`.
+
+How should these routes be merged? One invariant might be that the consumer of a
+service shouldn't be able to override the routes from the service owner.
+
+The route that applies to all requests:
+
+```yaml
+metadata:
+  name: owner
+spec:
+ targetRef:
+   kind: Mesh
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    default:
+      rules:
+      - matches:
+          path:
+            prefix: /
+        filters:
+        - responseHeader:
+            add:
+            - name: X-Some-Header
+              value: something
+        backendRefs:
+        - weight: 90
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v1
+        - weight: 10
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v2
+```
+
+The route that applies to requests from a specific service:
+
+```yaml
+metadata:
+  name: consumer
+spec:
+ targetRef:
+   kind: MeshService
+   name: svc_b
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    default:
+      rules:
+      - matches:
+          path:
+            prefix: /
+        filters:
+        - requestHeader:
+            add:
+            - name: X-Client
+              value:
+        backendRefs:
+        - weight: 100
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v1
+```
+
+when these resources are merged the latter will have higher priority and
+override the less specific resource:
+
+```yaml
+metadata:
+  name: consumer
+spec:
+ targetRef:
+   kind: MeshService
+   name: svc_b
+ to:
+  - targetRef:
+      kind: MeshService
+      name: svc_a
+    default:
+      rules:
+      - matches:
+          path:
+            prefix: /
+        filters:
+        - requestHeader:
+            add:
+            - name: X-Client
+              value:
+        backendRefs:
+        - weight: 100
+          kind: MeshServiceSubset
+          name: svc_a
+          tags:
+            version: v1
+```
+
+But the service owner probably expects to somehow
+be able to "enforce" routes on consumers that can't be overriden.
+
+Perhaps we finally need [`override`](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#hierarchy)?
```

Comment:
Yes, the other option would be to support `targetRef: MeshService(svc_a) + from: Mesh`. It's a route on the producer side, this way it's always executed regardless of what was set in consumer routes.

Reply:
Do you mean it would also get applied when receiving the traffic (on the producer DPP)?

Reply:
True that would probably work too, though I think we otherwise have an assumption that `targetRef` targets the resource whose config is being changed, which this would break. It's also not obvious to me that the merging logic still make sense if we starting merging `targetRef + to` and `from + targetRef`

Reply:
> assumption that targetRef targets the resource whose config is being changed

This is how it works in other policies, it's clear and I would be very much for not changing that.

Reply:
I'd say we decided to not do this for now.